---
title: "CTDPathSim"
author: "Banabithi Bose, Matt Moravec,  Serdar Bozdag"
date: "5/12/2020"
output:      
    BiocStyle::html_document:
        toc: true
        toc_depth: 3  
        number_sections: true
        theme: united 
r_packages: rmarkdown
vignette: > 
    %\VignetteIndexEntry{CTDPathSim} 
    %\VignetteEngine{knitr::rmarkdown} 
    %\VignetteEncoding[UTF-8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# <a id = "Introduction"></a> Introduction
CTDPathSim is a computational tool which utilizes a pathway activity-based approach to compute similarity scores between primary tumor samples and cell lines. The latest version CTDPathSim2.0 integrates DNA methylation, gene expression and copy number variation datasets. The package is able to run without copy number variation to produce the results based on previous version of CTDPathSim. CTDPathSim2.0 has five main computational steps:

1:Computing sample-specific deconvoluted DNA methylation profile
 
2:Computing sample-specific deconvoluted expression profile

3:Computing sample and cell line-specific differentially expressed (DE) genes and enriched biological pathways

4:Computing sample and cell line-specific differentially methylated (DM) and differentially aberrated (DA) genes

5:Computing sample-cell line pathway activity-based similarity score

This tool can be run without copy number data as well, that will produce the result as in our previous version of CTDPathSim.
```{r}
library(CTDPathSim2)
data("CTDData")
```

# <a id = "Example-Data-Sets"> </a> Example Data Sets
The example datasets are provided in this vignette to run CTDPathSim.

## <a id = "Reference.Methylation.Markers"> </a> Reference.Methylation.Markers
This file is a vector containing DNA methylation markers(important probes)of different cell types that are used in deconvolution purpose.
```{r}
Reference.Methylation.Markers[1:5]
```

## <a id = "Methylation.Sample.Rda"> </a> Methylation.Sample.Rda
This is a R dataframe object where the columns are the samples and the rows are the methylation probes.
```{r}
Methylation.Sample[1:5,1:4]
```

## <a id = "CancerGeneList.Rda"> </a> CancerGeneList.Rda
This R dataframe object has a single column containing cancer-driver genes.
```{r}
CancerGeneList[1:5,]
```
## <a id = "Expression.Sample.Rda"> </a> Expression.Sample.Rda
This R dataframe object has gene expression data for patient samples. The columns are the samples and the genes are in the rows. The user can use RPKM or FPKM normalized data. The tool has not been tested on count data.

```{r}
Expression.Sample[,1:4]
```

## <a id = "CNV.Sample.Rda"> </a> CNV.Sample.Rda
This R dataframe object has gene-centric copy number variation for the samples. The rows are the genes and columns are the samples.
```{r}
CNV.Sample[1:5,1:3]
```

## <a id = "Annotation450K.Rda"> </a> Annotation450K.Rda
This is a R dataframe/SummarizedExperiment object that includes methylation probe id's, and the corresponding reference gene name and reference group name. The columns in the dataframe/SummarizedExperiment have to be labeled to follow the convention of the annotation file in "Genome Browser" (https://genome.ucsc.edu/) as "ID, UCSC_REFGENE_NAME, UCSC_REFGENE_GROUP".

```{r}
Annotation450K[c(6,7,10,11),]
```

## <a id = "Annotation27K.Rda"> </a> Annotation27K.Rda

This is a R dataframe/SummarizedExperiment object that includes methylation probe id's, and the corresponding reference gene name and synonyms. The columns are labeled to follow the convention of the annotation file in "Genome Browser" as "ID, Symbol, Synonym".

```{r}
Annotation27K[1:5,]
```


## <a id = "Reference.Methylation.CellTypes.Rda"> </a> Reference.Methylation.CellTypes.Rda
This R dataframe/SummarizedExperiment object includes reference methylation profile of reference cell types needed for deconvolution. The rows are methylation probes and columns are cell type names.
```{r}
Reference.Methylation.CellTypes[1:5,1:5]
```

## <a id = "Reference.CellTypes.Names.Rda"> </a> Reference.CellTypes.Names.Rda
This R dataframe/SummarizedExperiment object has two columns containing the replicates and classes of the reference cell types used for deconvolution.

```{r}
Reference.CellTypes.Names[1:5,]
```


# <a id = "Using-the-Functions"> </a> Using the Functions
The following sections show the use of each function in the package to complete the pipeline.

## <a id = "Step-1"> </a> Step 1
Step one is computing sample-specific deconvoluted methylation profile. This step requires the use of the four functions, namely, *RunDeconvMethylFun()*, *PlotDeconvMethylFun()*, *ProbeToGeneFun()*, and *SampleDeconvProfileFun()*. 

### <a id = "RunDeconvMethylFun()"> </a> RunDeconvMethylFun()
The function RunDeconvMethyl(), computes the deconvoluted methylation profile and proportions of different cell types. This function has two arguments, *Example.Methylation*, *Reference.Methylation.Probes*. The *Example.Methylation* argument has all the methylation values for the samples and probes, and the argument *Reference.Methylation.Probes* includes all the marker probes. 

Methylation.Sample - A dataframe/SummarizedExperiment for the methylation values for with samples and probes.

Reference.Methylation.Probes - A vector of the probes(markers).


```{r results = "hide"}
stage1_result_ct <-
  RunDeconvMethylFun(Methylation.Sample = Methylation.Sample,
                     Reference.Methylation.Probes = Reference.Methylation.Markers)
```
The function RunDeconvMethyl() returns a list of seven objects. These objects are the deconvoluted methylation profile of the samples, the proportions of each cell type, the iterations, the explained variance, the residual sum of squares, the Akaike information criterion (AIC), and the residual sum of squares per iteration in the deconvolution algorithm.
```{r}
stage1_result_ct$methylation[1:5,]
```
```{r}
stage1_result_ct$proportions[1:5,]
```



### <a id = "PlotDeconvMethylFun()"> </a> PlotDeconvMethylFun()
PlotDeconvMethyl(), creates a correlation plot to show the correlation between the cell types used for deconvolution. User can use this plot to choose the appropriate cell type labels for the deconvoluted methylation profiles of samples.

Methylation.Sample - A R dataframe/SummarizedExperiment/SummarizedExperiment object where the columns are the samples and the rows are the methylation probes

Reference.Methylation.Markers - A vector with DNA methylation markers (i.e. most variable probes)of reference cell types 

Reference.Methylation.CellTypes - A R dataframe/SummarizedExperiment object with DNA methylation of reference cell types

stage1_result_ct - A list that contains the output of runDeconvMethyl

Reference.CellTypes.Names - A R dataframe/SummarizedExperiment object with the reference cell types and their classes

```{r}
print(
  PlotDeconvMethylFun(
    Methylation.Sample = Methylation.Sample,
    Reference.Methylation.Probes = Reference.Methylation.Markers,
    Reference.Methylation.CellTypes = Reference.Methylation.CellTypes,
    stage1_result_ct = stage1_result_ct,
    Reference.CellTypes.Names = Reference.CellTypes.Names
  )
)
dev.off()
```


### <a id = "ProbeTogeneFun()"> </a> ProbeTogeneFun()
ProbeTogene450Fun(), converts the methylation data from DNA methylation probes to genes. This function works for both 450K and 27K probe formats. When using this function user must indicates the format with the *geneFormat* argument which takes a character string of either "450K" or "27K". 
Methylation.Probe.Annotation - A R dataframe/SummarizedExperiment object that includes the probes, the symbol and synonyms according to Genome Browser(https://genome.ucsc.edu/)

Deconv_methylation - A R dataframe/SummarizedExperiment with deconvoluted DNA methylation probes of reference cell types that is generated from the output of runDeconvMethyl in stage1_result_ct object

Probe.Annotation - A character string to select "450K" or "27K" probes


```{r}
Deconv_methylation <- stage1_result_ct$methylation
Deconv_meth_gene <-
  ProbeToGeneFun(
    Methylation.Probe.Annotation = Annotation450K,
    Deconv_methylation = Deconv_methylation,
    Probe.Annotation = "450K"
  )
```

The output of this function is a R dataframe object with gene-centric DNA methylation (deconvoluted) of cell types. The rows are the genes and the columns are the cell types. The columns are labeled as V1, V2 and V3 etc. that represent the chosen cell types using *PlotDeconvMethylFun()*.

```{r}
Deconv_meth_gene[1:5,]
```


### <a id = "SampleDeconvProfileFun()-for-sample-specific-deconvoluted-methylation-profile"> </a> SampleDeconvProfileFun() for sample-specific deconvoluted methylation profile
This function makes sample-specific deconvoluted methylation profile for each sample. This function will save the files in folder inside the *CTDDirectory* labeled either "Sample_expression" or "Sample_methylation". This function runs with deconvoluted DNA methylation (gene-centric) that is generated by using RunDeconvMethylFun() and ProbeTogeneFun().The same can run with deconvoluted gene expression of samples that is generated by RunDeconvExprFun(). 

Deconv_cell_profile - A R dataframe/SummarizedExperiment object with deconvoluted DNA methylation generated using RunDeconvMethylFun() and ProbeTogeneFun()or deconvoluted expression of cell types generated by RunDeconvExprFun()

Deconv_proportions - A R dataframe/SummarizedExperiment object with sample-centric proportion of each cell type.
Rows are the samples, columns are the cell types generated from RunDeconvMethylFun()

ProfileType - A character string to indicate "Methylation" or "Expression" depending on Deconv_cell_profile

CTDDirectory - A character string for the file path of the directory. It should be the same for all functions

```{r}
CTDDirectory <- tempdir()
SampleDeconvProfileFun(
  Deconv_meth_gene,
  Deconv_proportions,
  ProfileType = "Methylation",
  CTDDirectory = CTDDirectory
)
```

The output for this function are the files that with deconvoluted DNA methylation profile of sample in CTDDirectory/sample_methylation. The rows are genes and columns are the cell types of each sample. 

```{r}
load("~/Sample_methylation/TCGA-OR-A5J1-01A.Rda")
methylation[1:5,]
```


## <a id = "Step-2"> </a> Step 2
In step 2, one computes the deconvoluted expression profile for each sample. Firstly, one computes the cell type-specific expression profiles for reference immune cells utilizing *RunDeconvExprFun()*.  Then, using cell type-specific proportions and cell type-specific expression as arguments, *SampleDeconvProfileFun()* infers the sample-specific deconvoluted expression profile of each tumor sample.

### <a id = "RunDeconvExprFun()"> </a> RunDeconvExprFun()
RunDeconvExpr(), creates a deconvoluted expression profile. In this function, one uses previously computed cell-type proportions for each sample from step 1 and the gene expression data of bulk tumor samples. This function, applies EDec stage 2 to compute cell-type specific expression profiles.

Expression.Sample - A R dataframe/SummarizedExperiment object with the expression values for each probe and cell type

Deconv_proportions - A R dataframe/SummarizedExperiment object with the proportion of each cell type for each patient

```{r}
Deconv_expression<-RunDeconvExprFun(Expression.Sample,Deconv_proportions)
```

The output of this function is an expression profile.
```{r}
Deconv_expression
```


### <a id = "SampleDeconvProfileFun()-for-sample-specific-deconvoluted-expression-profile"> </a> SampleDeconvProfileFun() for sample-specific deconvoluted expression profile

SampleDeconvProfileFun() is described in SampleDeconvProfileFun() Part 1, this section is included to demonstrate how this function is to be used with expression data in the CTDPathSim pipeline.

```{r}
SampleDeconvProfileFun(
  Deconv_expression,
  Deconv_proportions,
  ProfileType = "Expression",
  CTDDirectory = CTDDirectory
)
```

The output for this function are the files that with deconvoluted expression profile of sample in CTDDirectory/sample_expression. The rows are genes and columns are the cell types of each sample. 

```{r}
load("~/Sample_expression/TCGA-OR-A5J1-01A.Rda")
expression
```


## <a id = "Step-3"> </a> Step 3
In this step, the functions are designed to compute the enriched biological pathways listed in REACTOME [39] database for each sample and cell line. In this step *GetSampleDGeneFun()* computes sample-specific and/or cell line-specific DE genes and *GetPathFun()* identifies enriched biological pathways utilizing these DE genes for each sample and/or cell line. 

### <a id = "GetSampleDGenesFun()-Get-Sample-DE-Genes"> </a> GetSampleDGenesFun() Get Sample DE Genes
GetSampleDGeneFun(), is designed to find the median expression (or methylation) of a gene for a patient or cell line. User can use the argument *DType* by setting "DE" or "DM" to compute DE or DM genes, respectively. The result is stored in a directory named Patient_DE_genes as an individual data frame for each patient. The directory Patient_DE_genes will be a sub-directory in *CTDDirectory*.

Omics.Sample - R dataframe/SummarizedExperiment object with the expression or methylation values for each probe and cell type

ncore - An integer value for the number of cores (default = 1)

DType - A character string to indicate if the user is finding "DE" genes or "DM" genes

CTDDirectory - A character string for the file path of the directory. It should be the same for all functions
```{r}
GetSampleDGenesFun(Expression.Sample = Expression.Sample, DType = "DE", CTDDirectory = CTDDirectory)
```

The output of this function are the R dataframe objects for each sample with DE genes created in folder CTDDirectory/Patient_DE_genes/DE_Gene

```{r}
load("~/Patient_DE_genes/DE_Gene/TCGA-OR-A5J6-01A_pat_de_genes.Rda")
pat_de_genes
```


### <a id = "GetPathFun()"> </a> GetPathFun()
GetPathFun(), finds the pathway between cancer genes and specific patient DE genes. This function finds enriched biological pathways listed in REACTOME database via pathway enrichment. We utilized the known frequently mutated genes that are considered as cancer driver from the DE gene list for this analysis. User can skip this and use all the DE genes.

CancerGeneList - A R dataframe/SummarizedExperiment object with special genes (can be cancer-driving,all genes,or any gene list of choice)

pat_de_genes - A R dataframe/SummarizedExperiment object with the patients DE genes

```{r}
x<-GetPathFun(CancerGeneList,pat_de_genes)
```

The output of this function is an R dataframe object that includes three columns, "ID", "reactome_pathway", and "p.adjust.pat". A dataframe object is created for each sample. 

```{r}
x[1:5,]
```


## <a id = "Step-4"> </a> Step 4
In this step, the functions are designed to compute the sample and cell line-specific DM and DA genes. This step utilizes the functions *GetSampleDGeneFun()* and *GetCellLineDAFun()*.

### <a id = "GetSampleDGeneFun()-Get-Sample-DM-Genes"> </a> GetSampleDGeneFun() Get Sample DM Genes
GetSampleDGeneFun() is described in GetSampleDGeneFun() Part 1. This section is included to demomstrate how this function can be used for DNA methylation data in the CTDPathSim pipeline.

```{r}
GetSampleDGenesFun(
  Expression.Sample = Methylation.Sample,
  DType = "DM",
  CTDDirectory = CTDDirectory
)
```
The output of this function are the R dataframe objects for each sample with DM genes created in folder CTDDirectory/Patient_DM_genes/DM_Gene
```{r}
load("~/Patient_DM_genes/DM_Gene/TCGA-P6-A5OG-01A_pat_de_genes.Rda")
pat_de_genes
```


### <a id = "GetCellLineDAFun()"> </a> GetCellLineDAFun()
GetCellLineDAFun(), is designed to find the differentially aberrated genes in the cell lines. 

CNV.CellLine - R dataframe/SummarizedExperiment object with the copy number values for each probe and cell type

ncore - An integer value for the number of cores (default = 1)

CTDDirectory - A character string for the file path of the directory. It should be the same for all functions
```{r}
GetCellLineDAFun(CNV.Sample = CNV.Sample, CTDDirectory = CTDDirectory)
```

The output of this function are the R dataframe objects for each sample with DA genes created in folder CTDDirectory/CCLE_DA_genes/DA_Gene

```{r}
load("~/CCLE_DA_genes/DE_Gene/ABC1_LUNG_pat_de_genes.Rda")
cell_da_genes
```


## <a id = "Step-5"> </a> Step 5
This step is designed to compute a similarity score between the cell line and sample. The function in this step, *FindSimFun()* can be used for methylation, expression and copy number data.

### <a id = "FindSimFun()"> </a> FindSimFun()
 This function can be used to compute the Spearman rank correlation between each sample and cell line using expression, methylation, or copy number profiles.

Patient.D.Genes - A R dataframe/SummarizedExperiment object of a sample's DE/DM/DA genes.

Patient.Reactome - A R dataframe/SummarizedExperiment object of a sample's enriched pathways.

Patient.Expr.Meth.CNV - A R dataframe object with a sample's expression, methylation or copy number profile. Rows are the genes and column is the sample.

Cell.D.Genes - A R dataframe/SummarizedExperiment object of a cell lines's DE/DM/DA genes.

Cell.Reactome - A R dataframe/SummarizedExperiment object of a cell lines's enriched pathways.

Cell.Expr.Meth.CNV - A R dataframe/SummarizedExperiment object with a cell lines's expression, methylation or copy number profile. Rows are the genes and column is the cell line.

```{r}
data("CTDData")
patient1.cellLine1.spearman <-
  FindSimFun(
    Patient.D.Genes = pat_de_genes,
    Patient.Reactome = pat_reactome,
    Patient.Expr.Meth.CNV = expression,
    Cell.D.Genes = cell_de_genes,
    Cell.Reactome = cell_reactome,
    Cell.Expr.Meth.CNV = ccle_expr
  )
```

The output of this function is a R dataframe object with the columns, "spearman", "pat_de", "cell_de", "union_de", "reactome_genes", "de_in_reactome". Spearman is the spearman correlation rank. Pat_de is the number of DE genes the sample has. Cell_de is the number of DE genes the cell line has. Union_de is the number of DE genes the sample and cell line have in common. Reactome_genes is the number of genes in the reactome. De_in_reactome is the number of DE genes in the reactome.

```{r}
patient1.cellLine1.spearman
```


# <a id = "Summary"> </a> Summary
This is a five step process designed to create a similarity network between the patient sample and the patient cell line. The steps are:Computing sample-specific deconvoluted DNA methylation profile, computing sample-specific deconvoluted expression profile, computing sample and cell line-specific differentially expressed (DE) genes and enriched biological pathways, computing sample and cell line-specific differentially methylated (DM) and differentially aberrated (DA) genes and computing sample-cell line pathway activity-based similarity score.

# <a id = "Acknowledgments"> </a> Acknowledgments
1.This material is based upon work supported by the National Science Foundation under Grant #1950826
2.This work was supported by the National Institute of General Medical Sciences of the National Institutes of Health under Award Number R35GM133657


